style.md

```diff
@@ -1515,6 +1516,118 @@ Note that the compiler will not generate errors when using predeclared
 identifiers, but tools such as `go vet` should correctly point out these and
 other cases of shadowing.
 
+### Avoid `init()`
+
+Avoid `init()` where possible. At a minimum, all code within `init()` should:
+
+1. Be completely deterministic, regardless of program environment or invocation.
+2. Be completely self-contained within the program.
+3. Not depend on side-effects of other `init()` functions.
+4. Be able to run in any order. While `init()` ordering is well-known, code can
+   change, and thus `init()` functions that rely on specific sequencing make
+   code brittle and error-prone.
+5. Not access or manipulate global or environment state, such as machine
+   information, environment variables, working directory, program
+   arguments/inputs, etc.
+6. Not perform I/O, including both filesystem, network, and system calls.
+
+Code that cannot satisfy these requirements likely belongs as a helper to be
+called as part of `main()` (or elsewhere in a program's lifecycle), or be
+written as part of `main()` itself. In particular, libraries that are intended
+to be used by other programs should take special care to be completely
+deterministic and not perform "init magic".
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Foo struct {
+    // ...
+}
+
+var _defaultFoo Foo
+
+func init() {
+    _defaultFoo = Foo{
+        // ...
+    }
+}
+```
+
+</td><td>
+
+```go
+var _defaultFoo = Foo{
+    // ...
+}
+```
+
+</td></tr>
+<tr><td>
+
+```go
+type Config struct {
+    // ...
+}
+
+var _config Config
+
+func init() {
+    // Bad: based on current directory
+    cwd, _ := os.Getwd()
+
+    // Bad: I/O
+    raw, _ := ioutil.ReadFile(
+        path.Join(cwd, "config", "config.yaml"),
+    )
+
+    yaml.Unmarshal(raw, &_config)
+}
+```
+
+</td><td>
+
+```go
+type Config struct {
+    // ...
+}
+
+func loadConfig() Config {
+    cwd, _ := os.Getwd()
+    raw, _ := ioutil.ReadFile(
+        path.Join(cwd, "config", "config.yaml"),
+    )
+
+    var config Config
+    yaml.Unmarshal(raw, &config)
+
+    return config
+}
+```
+
+</td></tr>
+</tbody></table>
+
+Notable exceptions to `init()` determinism guidance include situations such as
+seeding non-cryptographic random number generators used across a codebase:
```

Comment:
Are we sure about this exception? It may be preferable for the `*rand.Rand` to be passed around explicitly, especially so that we can get determinism in tests of randomness.

Reply:
Good point about tests - I'll just omit this, I don't think there's a ton of value to delving too deep here.